# Q-Former v3 vs Concat Baseline: 3-Fold Cross-Validation

This notebook compares two architectures using the **v3 training improvements**:
1. **Q-Former v3**: Uses Q-Former fusion with all training improvements (differential LR, warmup, Huber loss, EMA, multi-task, R-Drop, RNA mask)
2. **Concat Baseline v3**: Bypasses Q-Former (mean pool), same training improvements

Both use the same pretrained checkpoint and identical training configuration for fair comparison.
Uses cell-line-aware 3-fold cross-validation with comprehensive statistical measures.

In [18]:
import sys
sys.path.insert(0, '..')

import torch
import torch.nn as nn
from torch.optim import AdamW
from torch.optim.lr_scheduler import CosineAnnealingLR, SequentialLR, LinearLR
from torch.amp import GradScaler, autocast
from torch.utils.data import DataLoader, Subset
from tqdm import tqdm
import numpy as np
from scipy import stats
from collections import defaultdict
import json
from pathlib import Path
import copy

from gastro_transformer.config import GastroTransformerConfig
from gastro_transformer.model import ModalitySlotQFormer
from gastro_transformer.data import (
    PairedMultiModalDataset,
    UnpairedModalityDataset,
    DrugEmbeddingDataset,
    IC50Dataset,
)
from gastro_transformer.losses import compute_ic50_metrics

In [20]:
# Configuration
DEVICE = 'cuda:1'
BATCH_SIZE = 256
N_FOLDS = 3
FINETUNE_EPOCHS = 10
SEED = 42

ROOT_DIR = '../'

# Both architectures use the same pretrained checkpoint
CHECKPOINT_PATH = ROOT_DIR + 'saved_checkpoints/pretrained_clrna.pt'

OUTPUT_DIR = Path(ROOT_DIR + 'reports/cross_validation_v3')
OUTPUT_DIR.mkdir(parents=True, exist_ok=True)

# Data paths
PAIRED_IMAGE_CSV = ROOT_DIR + 'data/processed/paired_image_ms-bcpp.csv'
PAIRED_RNA_CSV = ROOT_DIR + 'data/processed/paired_rna_ms-bcpp.csv'
UNPAIRED_IMAGE_CSV = ROOT_DIR + 'data/processed/unpaired_image.csv'
UNPAIRED_RNA_CSV = ROOT_DIR + 'data/processed/unpaired_rna.csv'
DRUG_EMBEDDINGS_CSV = ROOT_DIR + 'data/drug_embeddings.csv'
IC50_CSV = ROOT_DIR + 'data/ic50_data.csv'
CELLLINE_RNA_CSV = ROOT_DIR + 'data/processed/ccle_rna_for_ic50.csv'

# V3 training improvement settings
QFORMER_LR_RATIO = 0.2       # Pretrained components get this fraction of base LR
HUBER_DELTA = 1.5             # Huber loss delta
EMA_DECAY = 0.999             # EMA decay rate
LAMBDA_TISSUE = 0.1           # Multi-task tissue loss weight
RDROP_ALPHA = 0.5             # R-Drop consistency weight

print(f"Device: {DEVICE}")
print(f"Batch size: {BATCH_SIZE}")
print(f"Finetune epochs: {FINETUNE_EPOCHS}")
print(f"N folds: {N_FOLDS}")
print(f"Checkpoint: {CHECKPOINT_PATH}")
print(f"\nV3 improvements: Huber(delta={HUBER_DELTA}), EMA(decay={EMA_DECAY}), "
      f"Multi-task(lambda={LAMBDA_TISSUE}), R-Drop(alpha={RDROP_ALPHA}), "
      f"Differential LR(ratio={QFORMER_LR_RATIO}), Warmup, RNA mask")

Device: cuda:1
Batch size: 256
Finetune epochs: 10
N folds: 3
Checkpoint: ../saved_checkpoints/pretrained_clrna.pt

V3 improvements: Huber(delta=1.5), EMA(decay=0.999), Multi-task(lambda=0.1), R-Drop(alpha=0.5), Differential LR(ratio=0.2), Warmup, RNA mask


In [19]:
# Create config
config = GastroTransformerConfig()
config.paired_image_csv = PAIRED_IMAGE_CSV
config.paired_rna_csv = PAIRED_RNA_CSV
config.unpaired_image_csv = UNPAIRED_IMAGE_CSV
config.unpaired_rna_csv = UNPAIRED_RNA_CSV
config.drug_embeddings_csv = DRUG_EMBEDDINGS_CSV
config.ic50_csv = IC50_CSV
config.cellline_rna_csv = CELLLINE_RNA_CSV
config.batch_size = BATCH_SIZE
config.num_workers = 4
config.persistent_workers = True
config.qformer_finetune_lr_ratio = QFORMER_LR_RATIO

config.num_query_tokens = 32
config.qformer_layers = 6

In [4]:
# Load data
print("Loading data...")

# Paired data
paired_dataset = PairedMultiModalDataset(PAIRED_IMAGE_CSV, PAIRED_RNA_CSV)
paired_loader = DataLoader(
    paired_dataset,
    batch_size=config.batch_size,
    shuffle=True,
    num_workers=config.num_workers,
    persistent_workers=config.persistent_workers
)
print(f"Loaded paired dataset: {len(paired_dataset)} samples")

# Unpaired data
unpaired_image_dataset = UnpairedModalityDataset(
    UNPAIRED_IMAGE_CSV,
    modality='image',
    embedding_dim=config.image_dim
)
unpaired_image_loader = DataLoader(
    unpaired_image_dataset,
    batch_size=config.batch_size,
    shuffle=True,
    num_workers=config.num_workers,
    persistent_workers=config.persistent_workers
)
print(f"Loaded unpaired image dataset: {len(unpaired_image_dataset)} samples")

unpaired_rna_dataset = UnpairedModalityDataset(
    UNPAIRED_RNA_CSV,
    modality='rna',
    embedding_dim=config.rna_dim
)
unpaired_rna_loader = DataLoader(
    unpaired_rna_dataset,
    batch_size=config.batch_size,
    shuffle=True,
    num_workers=config.num_workers,
    persistent_workers=config.persistent_workers
)
print(f"Loaded unpaired RNA dataset: {len(unpaired_rna_dataset)} samples")

# IC50 data
drug_dataset = DrugEmbeddingDataset(DRUG_EMBEDDINGS_CSV)
ic50_dataset = IC50Dataset(
    IC50_CSV,
    drug_dataset,
    rna_csv_path=CELLLINE_RNA_CSV,
    add_tissue_ids=True
)
print(f"Loaded IC50 dataset: {len(ic50_dataset)} samples")

Loading data...
Found 100 paired samples
Loaded paired dataset: 100 samples
Loaded 148089 image samples from ../data/processed/unpaired_image.csv
Loaded unpaired image dataset: 148089 samples
Loaded 10314 rna samples from ../data/processed/unpaired_rna.csv
Loaded unpaired RNA dataset: 10314 samples
Loaded 234 drug embeddings from ../data/drug_embeddings.csv
Loaded 185294 IC50 entries from ../data/ic50_data.csv
After filtering to available drugs: 185294 entries
After removing duplicates: 173942 entries
IC50 values stored as-is (log_transform=False): range [-9.99, 12.90], std=2.774
Number of unique cell-lines: 998
Resolved tissue IDs for 998 unique cell-lines


../gastro_transformer/data.py:457: UserWarning: Found 11352 duplicate drug-cellline pairs. Keeping first occurrence of each pair.
  warnings.warn(


Loaded RNA embeddings for 592 cell-lines
RNA availability: 592/998 cell-lines have RNA, 406 missing
Loaded IC50 dataset: 173942 samples


In [5]:
# EMA (Exponential Moving Average) helper classes
class EMAModel:
    """Maintains shadow copy of weights: ema = decay * ema + (1-decay) * param."""

    def __init__(self, model, decay=0.999):
        self.decay = decay
        self.shadow = {}
        self.backup = {}
        for name, param in model.named_parameters():
            if param.requires_grad:
                self.shadow[name] = param.data.clone()

    @torch.no_grad()
    def update(self, model):
        for name, param in model.named_parameters():
            if param.requires_grad and name in self.shadow:
                self.shadow[name].mul_(self.decay).add_(param.data, alpha=1.0 - self.decay)

    def apply(self, model):
        return _EMAContext(self, model)


class _EMAContext:
    """Context manager: temporarily swap in EMA weights."""

    def __init__(self, ema, model):
        self.ema = ema
        self.model = model

    def __enter__(self):
        self.ema.backup = {}
        for name, param in self.model.named_parameters():
            if param.requires_grad and name in self.ema.shadow:
                self.ema.backup[name] = param.data.clone()
                param.data.copy_(self.ema.shadow[name])
        return self.model

    def __exit__(self, *args):
        for name, param in self.model.named_parameters():
            if name in self.ema.backup:
                param.data.copy_(self.ema.backup[name])
        self.ema.backup = {}


def build_param_groups(model, config):
    """Build parameter groups with differential learning rates.
    
    Pretrained components (Q-Former, image/rna projectors, image/rna type embeds) get low LR.
    New components (drug projector, drug/cellline type embeds, IC50 head, CellLineEncoder) get full LR.
    """
    base_lr = config.learning_rate
    low_lr = base_lr * config.qformer_finetune_lr_ratio
    
    param_groups = []
    assigned = set()
    
    # Q-Former: low LR (pretrained)
    qf_params = []
    for name, param in model.named_parameters():
        if param.requires_grad and 'qformer' in name:
            qf_params.append(param)
            assigned.add(name)
    if qf_params:
        param_groups.append({'params': qf_params, 'lr': low_lr})
    
    # Pretrained projectors (image, rna): low LR; Drug projector: full LR
    pretrained_proj, new_proj = [], []
    for name, param in model.named_parameters():
        if param.requires_grad and name not in assigned and 'projectors' in name:
            if 'projectors.image' in name or 'projectors.rna' in name:
                pretrained_proj.append(param)
            else:
                new_proj.append(param)
            assigned.add(name)
    if pretrained_proj:
        param_groups.append({'params': pretrained_proj, 'lr': low_lr})
    if new_proj:
        param_groups.append({'params': new_proj, 'lr': base_lr})
    
    # Pretrained type embeds (image, rna): low LR; New (drug, cellline): full LR
    pretrained_te, new_te = [], []
    for name, param in model.named_parameters():
        if param.requires_grad and name not in assigned and 'modality_type_embeddings' in name:
            if 'image' in name or 'rna' in name:
                pretrained_te.append(param)
            else:
                new_te.append(param)
            assigned.add(name)
    if pretrained_te:
        param_groups.append({'params': pretrained_te, 'lr': low_lr})
    if new_te:
        param_groups.append({'params': new_te, 'lr': base_lr})
    
    # IC50 head + CellLineEncoder: full LR
    ic50_params = []
    for name, param in model.named_parameters():
        if param.requires_grad and name not in assigned and ('ic50_head' in name or 'cellline_encoder' in name):
            ic50_params.append(param)
            assigned.add(name)
    if ic50_params:
        param_groups.append({'params': ic50_params, 'lr': base_lr})
    
    # Remaining params: full LR
    other = [p for n, p in model.named_parameters() if p.requires_grad and n not in assigned]
    if other:
        param_groups.append({'params': other, 'lr': base_lr})
    
    return param_groups


print("EMA and param group helpers defined.")

EMA and param group helpers defined.


In [6]:
# Create cell-line-aware CV splits
def create_cv_splits(ic50_dataset, n_folds=3, seed=42):
    cellline_ids = ic50_dataset.cellline_ids
    unique_celllines = np.unique(cellline_ids)

    np.random.seed(seed)
    np.random.shuffle(unique_celllines)

    fold_size = len(unique_celllines) // n_folds
    folds = []
    for i in range(n_folds):
        if i < n_folds - 1:
            test_cl = unique_celllines[i * fold_size:(i + 1) * fold_size]
        else:
            test_cl = unique_celllines[i * fold_size:]

        train_val_cl = np.array([c for c in unique_celllines if c not in test_cl])
        val_size = len(train_val_cl) // 4
        np.random.seed(seed + i)
        np.random.shuffle(train_val_cl)
        val_cl = train_val_cl[:val_size]
        train_cl = train_val_cl[val_size:]

        train_idx = np.where(np.isin(cellline_ids, train_cl))[0].tolist()
        val_idx = np.where(np.isin(cellline_ids, val_cl))[0].tolist()
        test_idx = np.where(np.isin(cellline_ids, test_cl))[0].tolist()

        folds.append((train_idx, val_idx, test_idx))

    return folds

cv_splits = create_cv_splits(ic50_dataset, N_FOLDS, SEED)
print(f"Dataset size: {len(ic50_dataset)}")
for i, (train_idx, val_idx, test_idx) in enumerate(cv_splits):
    print(f"Fold {i+1}: Train={len(train_idx)}, Val={len(val_idx)}, Test={len(test_idx)}")

Dataset size: 173942
Fold 1: Train=85830, Val=29184, Test=58928
Fold 2: Train=87667, Val=29101, Test=57174
Fold 3: Train=87536, Val=28566, Test=57840


In [7]:
def train_fold(model, config, train_loader, val_loader, device, finetune_epochs=10):
    """Train a single fold with all v3 improvements."""
    scaler = GradScaler("cuda") if "cuda" in device else None
    device_type = "cuda" if "cuda" in device else "cpu"

    best_val_loss = float("inf")
    best_val_metrics = None
    best_state = None
    history = defaultdict(list)

    # Differential LR param groups
    param_groups = build_param_groups(model, config)
    optimizer = AdamW(param_groups, weight_decay=config.weight_decay)

    # Warmup + Cosine Decay scheduler
    warmup_epochs = min(2, max(1, finetune_epochs // 5))
    warmup_sched = LinearLR(optimizer, start_factor=0.1, end_factor=1.0, total_iters=warmup_epochs)
    cosine_sched = CosineAnnealingLR(optimizer, T_max=max(1, finetune_epochs - warmup_epochs))
    scheduler = SequentialLR(optimizer, schedulers=[warmup_sched, cosine_sched], milestones=[warmup_epochs])

    # EMA
    ema = EMAModel(model, decay=EMA_DECAY)

    # Huber loss
    ic50_loss_fn = lambda pred, tgt: nn.functional.huber_loss(pred, tgt, delta=HUBER_DELTA)

    for epoch in range(finetune_epochs):
        model.train()
        train_loss = 0
        train_preds = []
        train_targets = []

        pbar = tqdm(train_loader, desc=f"Train Epoch {epoch+1}/{finetune_epochs}")

        for batch in pbar:
            optimizer.zero_grad()

            drug_embeds = batch["drug_embed"].to(device)
            ic50_targets = batch["ic50"].to(device)
            cellline_ids = batch["cellline_id"].to(device)

            cancer_type_ids = batch.get("cancer_type_id")
            if cancer_type_ids is not None:
                cancer_type_ids = cancer_type_ids.to(device)

            tissue_ids_batch = batch.get("tissue_id")
            if tissue_ids_batch is not None:
                tissue_ids_batch = tissue_ids_batch.to(device)

            cellline_rna_embeds = batch.get("rna_embed")
            if cellline_rna_embeds is not None:
                cellline_rna_embeds = cellline_rna_embeds.to(device)

            rna_available = batch.get("rna_available")
            if rna_available is not None:
                rna_available = rna_available.to(device)

            fwd_kwargs = dict(
                drug_embeds=drug_embeds,
                cellline_ids=cellline_ids,
                cancer_type_ids=cancer_type_ids,
                tissue_ids=tissue_ids_batch,
                cellline_rna_embeds=cellline_rna_embeds,
                rna_available=rna_available,
            )

            with autocast(device_type, enabled=scaler is not None):
                outputs = model(**fwd_kwargs)

                # Primary loss: Huber
                loss = ic50_loss_fn(outputs["ic50_pred"], ic50_targets)

                # Multi-task: tissue classification auxiliary loss
                if tissue_ids_batch is not None:
                    tissue_logits = outputs.get("tissue_logits")
                    if tissue_logits is not None:
                        tissue_loss = nn.functional.cross_entropy(tissue_logits, tissue_ids_batch)
                        loss = loss + LAMBDA_TISSUE * tissue_loss

                # R-Drop: two forward passes, penalize divergence
                outputs2 = model(**fwd_kwargs)
                loss2 = ic50_loss_fn(outputs2["ic50_pred"], ic50_targets)
                rdrop_loss = nn.functional.mse_loss(outputs["ic50_pred"], outputs2["ic50_pred"])
                loss = 0.5 * (loss + loss2) + RDROP_ALPHA * rdrop_loss

            if scaler:
                scaler.scale(loss).backward()
                scaler.unscale_(optimizer)
                torch.nn.utils.clip_grad_norm_(model.parameters(), config.max_grad_norm)
                scaler.step(optimizer)
                scaler.update()
            else:
                loss.backward()
                torch.nn.utils.clip_grad_norm_(model.parameters(), config.max_grad_norm)
                optimizer.step()

            # EMA update
            ema.update(model)

            train_loss += loss.item()
            train_preds.extend(outputs["ic50_pred"].detach().cpu().numpy())
            train_targets.extend(ic50_targets.cpu().numpy())

            pbar.set_postfix({"loss": f"{train_loss / (pbar.n + 1):.4f}"})

        avg_train_loss = train_loss / len(train_loader)
        train_metrics = compute_ic50_metrics(
            torch.tensor(train_preds),
            torch.tensor(train_targets)
        )
        history["train_loss"].append(avg_train_loss)
        history["train_r2"].append(train_metrics["r2"])

        # Validation with EMA weights
        with ema.apply(model):
            val_loss, val_preds, val_tgts = _validate(model, val_loader, device, ic50_loss_fn)

        val_metrics = compute_ic50_metrics(torch.tensor(val_preds), torch.tensor(val_tgts))
        avg_val_loss = val_loss / max(len(val_loader), 1)
        history["val_loss"].append(avg_val_loss)
        history["val_r2"].append(val_metrics["r2"])

        print(f"Epoch {epoch+1}/{finetune_epochs} - Train Loss: {avg_train_loss:.4f}, "
              f"Train R2: {train_metrics['r2']:.4f}, Val Loss: {avg_val_loss:.4f}, "
              f"Val R2: {val_metrics['r2']:.4f}")

        if avg_val_loss < best_val_loss:
            best_val_loss = avg_val_loss
            best_val_metrics = val_metrics
            # Save EMA state dict as best
            with ema.apply(model):
                best_state = copy.deepcopy(model.state_dict())

        scheduler.step()

    # Restore best EMA weights for test evaluation
    if best_state is not None:
        model.load_state_dict(best_state)

    return {
        "history": dict(history),
        "best_val_loss": best_val_loss,
        "best_val_metrics": best_val_metrics,
    }


def _validate(model, loader, device, loss_fn):
    """Run validation, return (total_loss, predictions, targets)."""
    model.eval()
    total_loss = 0
    preds, targets = [], []

    with torch.no_grad():
        for batch in loader:
            drug_embeds = batch["drug_embed"].to(device)
            ic50_targets = batch["ic50"].to(device)
            cellline_ids = batch["cellline_id"].to(device)

            cancer_type_ids = batch.get("cancer_type_id")
            if cancer_type_ids is not None:
                cancer_type_ids = cancer_type_ids.to(device)
            tissue_ids_batch = batch.get("tissue_id")
            if tissue_ids_batch is not None:
                tissue_ids_batch = tissue_ids_batch.to(device)
            cellline_rna_embeds = batch.get("rna_embed")
            if cellline_rna_embeds is not None:
                cellline_rna_embeds = cellline_rna_embeds.to(device)
            rna_available = batch.get("rna_available")
            if rna_available is not None:
                rna_available = rna_available.to(device)

            outputs = model(
                drug_embeds=drug_embeds,
                cellline_ids=cellline_ids,
                cancer_type_ids=cancer_type_ids,
                tissue_ids=tissue_ids_batch,
                cellline_rna_embeds=cellline_rna_embeds,
                rna_available=rna_available,
            )
            loss = loss_fn(outputs["ic50_pred"], ic50_targets)
            total_loss += loss.item()
            preds.extend(outputs["ic50_pred"].cpu().numpy())
            targets.extend(ic50_targets.cpu().numpy())

    return total_loss, preds, targets

In [8]:
def evaluate_on_test(model, test_loader, device):
    """Evaluate model on test set (model should already have best weights loaded)."""
    model.eval()
    test_preds = []
    test_targets = []

    with torch.no_grad():
        for batch in test_loader:
            drug_embeds = batch["drug_embed"].to(device)
            ic50_targets = batch["ic50"].to(device)
            cellline_ids = batch["cellline_id"].to(device)

            cancer_type_ids = batch.get("cancer_type_id")
            if cancer_type_ids is not None:
                cancer_type_ids = cancer_type_ids.to(device)

            tissue_ids_batch = batch.get("tissue_id")
            if tissue_ids_batch is not None:
                tissue_ids_batch = tissue_ids_batch.to(device)

            cellline_rna_embeds = batch.get("rna_embed")
            if cellline_rna_embeds is not None:
                cellline_rna_embeds = cellline_rna_embeds.to(device)

            rna_available = batch.get("rna_available")
            if rna_available is not None:
                rna_available = rna_available.to(device)

            outputs = model(
                drug_embeds=drug_embeds,
                cellline_ids=cellline_ids,
                cancer_type_ids=cancer_type_ids,
                tissue_ids=tissue_ids_batch,
                cellline_rna_embeds=cellline_rna_embeds,
                rna_available=rna_available,
            )

            test_preds.extend(outputs["ic50_pred"].cpu().numpy())
            test_targets.extend(ic50_targets.cpu().numpy())

    test_metrics = compute_ic50_metrics(
        torch.tensor(test_preds),
        torch.tensor(test_targets)
    )

    return {
        "predictions": test_preds,
        "targets": test_targets,
        "metrics": test_metrics
    }

In [9]:
def run_cross_validation(ic50_dataset, cv_splits, config, use_qformer, device,
                         finetune_epochs=10, checkpoint_path=None):
    """Run cross-validation for a specific architecture with v3 training improvements."""
    arch_name = "Q-Former v3" if use_qformer else "Concat Baseline v3"

    print(f"\n{'='*60}")
    print(f"Running CV for: {arch_name}")
    print(f"use_qformer={use_qformer}, checkpoint={checkpoint_path}")
    print(f"Improvements: Huber, EMA, Multi-task, R-Drop, Differential LR, Warmup, RNA mask")
    print(f"{'='*60}")

    fold_metrics = []
    all_preds = []
    all_targets = []

    for fold_idx, (train_idx, val_idx, test_idx) in enumerate(cv_splits):
        print(f"\n--- Fold {fold_idx + 1}/{len(cv_splits)} ---")

        # Create data loaders
        train_dataset = Subset(ic50_dataset, train_idx)
        val_dataset = Subset(ic50_dataset, val_idx)
        test_dataset = Subset(ic50_dataset, test_idx)

        train_loader = DataLoader(
            train_dataset,
            batch_size=config.batch_size,
            shuffle=True,
            num_workers=config.num_workers,
            persistent_workers=config.persistent_workers
        )
        val_loader = DataLoader(
            val_dataset,
            batch_size=config.batch_size,
            shuffle=False,
            num_workers=config.num_workers,
            persistent_workers=config.persistent_workers
        )
        test_loader = DataLoader(
            test_dataset,
            batch_size=config.batch_size,
            shuffle=False,
            num_workers=config.num_workers,
            persistent_workers=config.persistent_workers
        )

        # Create model with appropriate Q-Former setting
        fold_config = copy.deepcopy(config)
        fold_config.use_qformer = use_qformer
        model = ModalitySlotQFormer(fold_config)
        model = model.to(device)

        # Load pretrained checkpoint
        if checkpoint_path and Path(checkpoint_path).exists():
            print(f"Loading checkpoint: {checkpoint_path}")
            checkpoint = torch.load(checkpoint_path, map_location=device, weights_only=False)
            model.load_state_dict(checkpoint["model_state_dict"], strict=False)

        # Train fold (returns model with best EMA weights loaded)
        train_fold(model, fold_config, train_loader, val_loader, device, finetune_epochs=finetune_epochs)

        # Evaluate on test set (model already has best weights)
        test_results = evaluate_on_test(model, test_loader, device)

        print(f"Fold {fold_idx + 1} Test - R2: {test_results['metrics']['r2']:.4f}, "
              f"Pearson R: {test_results['metrics']['pearson_r']:.4f}, "
              f"Spearman R: {test_results['metrics'].get('spearman_r', 0):.4f}, "
              f"RMSE: {test_results['metrics']['rmse']:.4f}, "
              f"MAE: {test_results['metrics']['mae']:.4f}")

        fold_metrics.append(test_results["metrics"])
        all_preds.extend(test_results["predictions"])
        all_targets.extend(test_results["targets"])

        # Clear GPU memory
        del model
        torch.cuda.empty_cache()

    return {
        "fold_metrics": fold_metrics,
        "all_preds": all_preds,
        "all_targets": all_targets
    }

In [12]:
# Run Q-Former v3 CV (with all training improvements)
qformer_results = run_cross_validation(
    ic50_dataset=ic50_dataset,
    cv_splits=cv_splits,
    config=config,
    use_qformer=True,
    device=DEVICE,
    finetune_epochs=FINETUNE_EPOCHS,
    checkpoint_path=CHECKPOINT_PATH
)


Running CV for: Q-Former v3
use_qformer=True, checkpoint=../saved_checkpoints/pretrained_clrna.pt
Improvements: Huber, EMA, Multi-task, R-Drop, Differential LR, Warmup, RNA mask

--- Fold 1/3 ---
Loading checkpoint: ../saved_checkpoints/pretrained_clrna.pt


Train Epoch 1/10: 100%|██████████████████████████████████████████████████████████████████████████████████████████| 336/336 [01:56<00:00,  2.88it/s, loss=1.8518]


Epoch 1/10 - Train Loss: 1.8518, Train R2: 0.2981, Val Loss: 2.8191, Val R2: -0.1985


Train Epoch 2/10: 100%|██████████████████████████████████████████████████████████████████████████████████████████| 336/336 [01:50<00:00,  3.04it/s, loss=0.9755]
/opt/conda/lib/python3.10/site-packages/torch/optim/lr_scheduler.py:198: UserWarning: The epoch parameter in `scheduler.step()` was not necessary and is being deprecated where possible. Please use `scheduler.step()` to step the scheduler. During the deprecation, if epoch is different from None, the closed form is used instead of the new chainable form, where available. Please open an issue if you are unable to replicate your use case: https://github.com/pytorch/pytorch/issues/new/choose.
  warnings.warn(EPOCH_DEPRECATION_WARNING, UserWarning)


Epoch 2/10 - Train Loss: 0.9755, Train R2: 0.7075, Val Loss: 2.2288, Val R2: 0.1046


Train Epoch 3/10: 100%|██████████████████████████████████████████████████████████████████████████████████████████| 336/336 [01:50<00:00,  3.04it/s, loss=0.7991]


Epoch 3/10 - Train Loss: 0.7991, Train R2: 0.7600, Val Loss: 1.7091, Val R2: 0.3566


Train Epoch 4/10: 100%|██████████████████████████████████████████████████████████████████████████████████████████| 336/336 [01:50<00:00,  3.05it/s, loss=0.7138]


Epoch 4/10 - Train Loss: 0.7138, Train R2: 0.7860, Val Loss: 1.2793, Val R2: 0.5595


Train Epoch 5/10: 100%|██████████████████████████████████████████████████████████████████████████████████████████| 336/336 [01:50<00:00,  3.05it/s, loss=0.6537]


Epoch 5/10 - Train Loss: 0.6537, Train R2: 0.8054, Val Loss: 0.9963, Val R2: 0.6778


Train Epoch 6/10: 100%|██████████████████████████████████████████████████████████████████████████████████████████| 336/336 [01:50<00:00,  3.05it/s, loss=0.5987]


Epoch 6/10 - Train Loss: 0.5987, Train R2: 0.8233, Val Loss: 0.8703, Val R2: 0.7221


Train Epoch 7/10: 100%|██████████████████████████████████████████████████████████████████████████████████████████| 336/336 [01:49<00:00,  3.07it/s, loss=0.5487]


Epoch 7/10 - Train Loss: 0.5487, Train R2: 0.8401, Val Loss: 0.8241, Val R2: 0.7363


Train Epoch 8/10: 100%|██████████████████████████████████████████████████████████████████████████████████████████| 336/336 [00:50<00:00,  6.72it/s, loss=0.5007]


Epoch 8/10 - Train Loss: 0.5007, Train R2: 0.8562, Val Loss: 0.8084, Val R2: 0.7402


Train Epoch 9/10: 100%|██████████████████████████████████████████████████████████████████████████████████████████| 336/336 [00:50<00:00,  6.70it/s, loss=0.4621]


Epoch 9/10 - Train Loss: 0.4621, Train R2: 0.8691, Val Loss: 0.8063, Val R2: 0.7398


Train Epoch 10/10: 100%|█████████████████████████████████████████████████████████████████████████████████████████| 336/336 [00:50<00:00,  6.70it/s, loss=0.4350]


Epoch 10/10 - Train Loss: 0.4350, Train R2: 0.8776, Val Loss: 0.8096, Val R2: 0.7378
Fold 1 Test - R2: 0.7393, Pearson R: 0.8605, Spearman R: 0.8322, RMSE: 1.4171, MAE: 1.0455

--- Fold 2/3 ---
Loading checkpoint: ../saved_checkpoints/pretrained_clrna.pt


Train Epoch 1/10: 100%|██████████████████████████████████████████████████████████████████████████████████████████| 343/343 [00:56<00:00,  6.06it/s, loss=1.8312]


Epoch 1/10 - Train Loss: 1.8312, Train R2: 0.3072, Val Loss: 2.7554, Val R2: -0.1514


Train Epoch 2/10: 100%|██████████████████████████████████████████████████████████████████████████████████████████| 343/343 [00:51<00:00,  6.66it/s, loss=0.9718]


Epoch 2/10 - Train Loss: 0.9718, Train R2: 0.7081, Val Loss: 2.1736, Val R2: 0.1357


Train Epoch 3/10: 100%|██████████████████████████████████████████████████████████████████████████████████████████| 343/343 [00:51<00:00,  6.67it/s, loss=0.7969]


Epoch 3/10 - Train Loss: 0.7969, Train R2: 0.7600, Val Loss: 1.6706, Val R2: 0.3724


Train Epoch 4/10: 100%|██████████████████████████████████████████████████████████████████████████████████████████| 343/343 [00:51<00:00,  6.67it/s, loss=0.7068]


Epoch 4/10 - Train Loss: 0.7068, Train R2: 0.7877, Val Loss: 1.2692, Val R2: 0.5597


Train Epoch 5/10: 100%|██████████████████████████████████████████████████████████████████████████████████████████| 343/343 [00:58<00:00,  5.87it/s, loss=0.6490]


Epoch 5/10 - Train Loss: 0.6490, Train R2: 0.8068, Val Loss: 1.0019, Val R2: 0.6728


Train Epoch 6/10: 100%|██████████████████████████████████████████████████████████████████████████████████████████| 343/343 [00:51<00:00,  6.68it/s, loss=0.5955]


Epoch 6/10 - Train Loss: 0.5955, Train R2: 0.8240, Val Loss: 0.8806, Val R2: 0.7160


Train Epoch 7/10: 100%|██████████████████████████████████████████████████████████████████████████████████████████| 343/343 [00:51<00:00,  6.69it/s, loss=0.5473]


Epoch 7/10 - Train Loss: 0.5473, Train R2: 0.8405, Val Loss: 0.8355, Val R2: 0.7297


Train Epoch 8/10: 100%|██████████████████████████████████████████████████████████████████████████████████████████| 343/343 [00:51<00:00,  6.69it/s, loss=0.4995]


Epoch 8/10 - Train Loss: 0.4995, Train R2: 0.8566, Val Loss: 0.8197, Val R2: 0.7337


Train Epoch 9/10: 100%|██████████████████████████████████████████████████████████████████████████████████████████| 343/343 [00:51<00:00,  6.69it/s, loss=0.4601]


Epoch 9/10 - Train Loss: 0.4601, Train R2: 0.8693, Val Loss: 0.8168, Val R2: 0.7336


Train Epoch 10/10: 100%|█████████████████████████████████████████████████████████████████████████████████████████| 343/343 [00:51<00:00,  6.70it/s, loss=0.4339]


Epoch 10/10 - Train Loss: 0.4339, Train R2: 0.8779, Val Loss: 0.8192, Val R2: 0.7319
Fold 2 Test - R2: 0.7385, Pearson R: 0.8606, Spearman R: 0.8283, RMSE: 1.4142, MAE: 1.0516

--- Fold 3/3 ---
Loading checkpoint: ../saved_checkpoints/pretrained_clrna.pt


Train Epoch 1/10: 100%|██████████████████████████████████████████████████████████████████████████████████████████| 342/342 [00:53<00:00,  6.41it/s, loss=1.8510]


Epoch 1/10 - Train Loss: 1.8510, Train R2: 0.2975, Val Loss: 2.7784, Val R2: -0.1722


Train Epoch 2/10: 100%|██████████████████████████████████████████████████████████████████████████████████████████| 342/342 [00:51<00:00,  6.68it/s, loss=0.9930]


Epoch 2/10 - Train Loss: 0.9930, Train R2: 0.6989, Val Loss: 2.1785, Val R2: 0.1280


Train Epoch 3/10: 100%|██████████████████████████████████████████████████████████████████████████████████████████| 342/342 [00:51<00:00,  6.69it/s, loss=0.8069]


Epoch 3/10 - Train Loss: 0.8069, Train R2: 0.7566, Val Loss: 1.6665, Val R2: 0.3703


Train Epoch 4/10: 100%|██████████████████████████████████████████████████████████████████████████████████████████| 342/342 [00:51<00:00,  6.68it/s, loss=0.7178]


Epoch 4/10 - Train Loss: 0.7178, Train R2: 0.7835, Val Loss: 1.2535, Val R2: 0.5618


Train Epoch 5/10: 100%|██████████████████████████████████████████████████████████████████████████████████████████| 342/342 [00:51<00:00,  6.67it/s, loss=0.6530]


Epoch 5/10 - Train Loss: 0.6530, Train R2: 0.8046, Val Loss: 0.9843, Val R2: 0.6765


Train Epoch 6/10: 100%|██████████████████████████████████████████████████████████████████████████████████████████| 342/342 [00:51<00:00,  6.67it/s, loss=0.5996]


Epoch 6/10 - Train Loss: 0.5996, Train R2: 0.8224, Val Loss: 0.8622, Val R2: 0.7210


Train Epoch 7/10: 100%|██████████████████████████████████████████████████████████████████████████████████████████| 342/342 [00:51<00:00,  6.70it/s, loss=0.5464]


Epoch 7/10 - Train Loss: 0.5464, Train R2: 0.8401, Val Loss: 0.8187, Val R2: 0.7349


Train Epoch 8/10: 100%|██████████████████████████████████████████████████████████████████████████████████████████| 342/342 [00:51<00:00,  6.69it/s, loss=0.5013]


Epoch 8/10 - Train Loss: 0.5013, Train R2: 0.8552, Val Loss: 0.8053, Val R2: 0.7383


Train Epoch 9/10: 100%|██████████████████████████████████████████████████████████████████████████████████████████| 342/342 [00:51<00:00,  6.69it/s, loss=0.4610]


Epoch 9/10 - Train Loss: 0.4610, Train R2: 0.8688, Val Loss: 0.8050, Val R2: 0.7373


Train Epoch 10/10: 100%|█████████████████████████████████████████████████████████████████████████████████████████| 342/342 [00:51<00:00,  6.70it/s, loss=0.4352]


Epoch 10/10 - Train Loss: 0.4352, Train R2: 0.8773, Val Loss: 0.8096, Val R2: 0.7347
Fold 3 Test - R2: 0.7481, Pearson R: 0.8652, Spearman R: 0.8331, RMSE: 1.3952, MAE: 1.0343


In [13]:
# Run Concat Baseline v3 CV (same improvements, no Q-Former — uses mean pooling)
concat_results = run_cross_validation(
    ic50_dataset=ic50_dataset,
    cv_splits=cv_splits,
    config=config,
    use_qformer=False,
    device=DEVICE,
    finetune_epochs=FINETUNE_EPOCHS,
    checkpoint_path=CHECKPOINT_PATH
)


Running CV for: Concat Baseline v3
use_qformer=False, checkpoint=../saved_checkpoints/pretrained_clrna.pt
Improvements: Huber, EMA, Multi-task, R-Drop, Differential LR, Warmup, RNA mask

--- Fold 1/3 ---
Loading checkpoint: ../saved_checkpoints/pretrained_clrna.pt


Train Epoch 1/10: 100%|██████████████████████████████████████████████████████████████████████████████████████████| 336/336 [00:05<00:00, 56.88it/s, loss=2.1139]


Epoch 1/10 - Train Loss: 2.0824, Train R2: 0.1996, Val Loss: 2.6648, Val R2: -0.1196


Train Epoch 2/10: 100%|██████████████████████████████████████████████████████████████████████████████████████████| 336/336 [00:05<00:00, 63.42it/s, loss=1.0999]


Epoch 2/10 - Train Loss: 1.0868, Train R2: 0.6844, Val Loss: 2.1829, Val R2: 0.1174


Train Epoch 3/10: 100%|██████████████████████████████████████████████████████████████████████████████████████████| 336/336 [00:04<00:00, 67.36it/s, loss=0.8540]


Epoch 3/10 - Train Loss: 0.8464, Train R2: 0.7588, Val Loss: 1.7947, Val R2: 0.2985


Train Epoch 4/10: 100%|██████████████████████████████████████████████████████████████████████████████████████████| 336/336 [00:04<00:00, 69.93it/s, loss=0.7607]


Epoch 4/10 - Train Loss: 0.7494, Train R2: 0.7798, Val Loss: 1.4534, Val R2: 0.4610


Train Epoch 5/10: 100%|██████████████████████████████████████████████████████████████████████████████████████████| 336/336 [00:04<00:00, 67.92it/s, loss=0.7109]


Epoch 5/10 - Train Loss: 0.6982, Train R2: 0.7934, Val Loss: 1.1651, Val R2: 0.5978


Train Epoch 6/10: 100%|██████████████████████████████████████████████████████████████████████████████████████████| 336/336 [00:04<00:00, 68.93it/s, loss=0.6659]


Epoch 6/10 - Train Loss: 0.6620, Train R2: 0.8042, Val Loss: 0.9708, Val R2: 0.6827


Train Epoch 7/10: 100%|██████████████████████████████████████████████████████████████████████████████████████████| 336/336 [00:05<00:00, 66.31it/s, loss=0.6297]


Epoch 7/10 - Train Loss: 0.6259, Train R2: 0.8158, Val Loss: 0.8694, Val R2: 0.7217


Train Epoch 8/10: 100%|██████████████████████████████████████████████████████████████████████████████████████████| 336/336 [00:04<00:00, 70.34it/s, loss=0.5961]


Epoch 8/10 - Train Loss: 0.5961, Train R2: 0.8259, Val Loss: 0.8264, Val R2: 0.7363


Train Epoch 9/10: 100%|██████████████████████████████████████████████████████████████████████████████████████████| 336/336 [00:04<00:00, 70.93it/s, loss=0.5775]


Epoch 9/10 - Train Loss: 0.5723, Train R2: 0.8337, Val Loss: 0.8103, Val R2: 0.7410


Train Epoch 10/10: 100%|█████████████████████████████████████████████████████████████████████████████████████████| 336/336 [00:04<00:00, 68.67it/s, loss=0.5579]


Epoch 10/10 - Train Loss: 0.5562, Train R2: 0.8387, Val Loss: 0.8049, Val R2: 0.7423
Fold 1 Test - R2: 0.7409, Pearson R: 0.8610, Spearman R: 0.8332, RMSE: 1.4126, MAE: 1.0475

--- Fold 2/3 ---
Loading checkpoint: ../saved_checkpoints/pretrained_clrna.pt


Train Epoch 1/10: 100%|██████████████████████████████████████████████████████████████████████████████████████████| 343/343 [00:05<00:00, 59.65it/s, loss=2.1126]


Epoch 1/10 - Train Loss: 2.0695, Train R2: 0.2038, Val Loss: 2.6320, Val R2: -0.0903


Train Epoch 2/10: 100%|██████████████████████████████████████████████████████████████████████████████████████████| 343/343 [00:05<00:00, 64.59it/s, loss=1.0916]


Epoch 2/10 - Train Loss: 1.0853, Train R2: 0.6823, Val Loss: 2.1721, Val R2: 0.1288


Train Epoch 3/10: 100%|██████████████████████████████████████████████████████████████████████████████████████████| 343/343 [00:05<00:00, 68.16it/s, loss=0.8525]


Epoch 3/10 - Train Loss: 0.8401, Train R2: 0.7593, Val Loss: 1.7943, Val R2: 0.3006


Train Epoch 4/10: 100%|██████████████████████████████████████████████████████████████████████████████████████████| 343/343 [00:04<00:00, 70.12it/s, loss=0.7468]


Epoch 4/10 - Train Loss: 0.7446, Train R2: 0.7803, Val Loss: 1.4573, Val R2: 0.4601


Train Epoch 5/10: 100%|██████████████████████████████████████████████████████████████████████████████████████████| 343/343 [00:04<00:00, 69.78it/s, loss=0.6987]


Epoch 5/10 - Train Loss: 0.6946, Train R2: 0.7941, Val Loss: 1.1672, Val R2: 0.5965


Train Epoch 6/10: 100%|██████████████████████████████████████████████████████████████████████████████████████████| 343/343 [00:05<00:00, 65.32it/s, loss=0.6648]


Epoch 6/10 - Train Loss: 0.6531, Train R2: 0.8067, Val Loss: 0.9742, Val R2: 0.6794


Train Epoch 7/10: 100%|██████████████████████████████████████████████████████████████████████████████████████████| 343/343 [00:05<00:00, 68.14it/s, loss=0.6182]


Epoch 7/10 - Train Loss: 0.6182, Train R2: 0.8183, Val Loss: 0.8805, Val R2: 0.7148


Train Epoch 8/10: 100%|██████████████████████████████████████████████████████████████████████████████████████████| 343/343 [00:05<00:00, 66.19it/s, loss=0.5930]


Epoch 8/10 - Train Loss: 0.5878, Train R2: 0.8282, Val Loss: 0.8424, Val R2: 0.7278


Train Epoch 9/10: 100%|██████████████████████████████████████████████████████████████████████████████████████████| 343/343 [00:05<00:00, 68.11it/s, loss=0.5639]


Epoch 9/10 - Train Loss: 0.5639, Train R2: 0.8364, Val Loss: 0.8276, Val R2: 0.7322


Train Epoch 10/10: 100%|█████████████████████████████████████████████████████████████████████████████████████████| 343/343 [00:05<00:00, 68.01it/s, loss=0.5525]


Epoch 10/10 - Train Loss: 0.5493, Train R2: 0.8411, Val Loss: 0.8237, Val R2: 0.7330
Fold 2 Test - R2: 0.7390, Pearson R: 0.8599, Spearman R: 0.8279, RMSE: 1.4130, MAE: 1.0531

--- Fold 3/3 ---
Loading checkpoint: ../saved_checkpoints/pretrained_clrna.pt


Train Epoch 1/10: 100%|██████████████████████████████████████████████████████████████████████████████████████████| 342/342 [00:05<00:00, 61.71it/s, loss=2.0927]


Epoch 1/10 - Train Loss: 2.0743, Train R2: 0.2023, Val Loss: 2.6425, Val R2: -0.1043


Train Epoch 2/10: 100%|██████████████████████████████████████████████████████████████████████████████████████████| 342/342 [00:05<00:00, 66.53it/s, loss=1.0814]


Epoch 2/10 - Train Loss: 1.0688, Train R2: 0.6903, Val Loss: 2.1619, Val R2: 0.1265


Train Epoch 3/10: 100%|██████████████████████████████████████████████████████████████████████████████████████████| 342/342 [00:04<00:00, 70.56it/s, loss=0.8495]


Epoch 3/10 - Train Loss: 0.8495, Train R2: 0.7559, Val Loss: 1.7702, Val R2: 0.3061


Train Epoch 4/10: 100%|██████████████████████████████████████████████████████████████████████████████████████████| 342/342 [00:05<00:00, 66.81it/s, loss=0.7580]


Epoch 4/10 - Train Loss: 0.7513, Train R2: 0.7780, Val Loss: 1.4183, Val R2: 0.4734


Train Epoch 5/10: 100%|██████████████████████████████████████████████████████████████████████████████████████████| 342/342 [00:05<00:00, 65.33it/s, loss=0.7023]


Epoch 5/10 - Train Loss: 0.6982, Train R2: 0.7923, Val Loss: 1.1257, Val R2: 0.6096


Train Epoch 6/10: 100%|██████████████████████████████████████████████████████████████████████████████████████████| 342/342 [00:05<00:00, 67.20it/s, loss=0.6696]


Epoch 6/10 - Train Loss: 0.6617, Train R2: 0.8031, Val Loss: 0.9455, Val R2: 0.6868


Train Epoch 7/10: 100%|██████████████████████████████████████████████████████████████████████████████████████████| 342/342 [00:05<00:00, 66.72it/s, loss=0.6255]


Epoch 7/10 - Train Loss: 0.6218, Train R2: 0.8164, Val Loss: 0.8600, Val R2: 0.7198


Train Epoch 8/10: 100%|██████████████████████████████████████████████████████████████████████████████████████████| 342/342 [00:05<00:00, 66.86it/s, loss=0.5895]


Epoch 8/10 - Train Loss: 0.5895, Train R2: 0.8270, Val Loss: 0.8266, Val R2: 0.7315


Train Epoch 9/10: 100%|██████████████████████████████████████████████████████████████████████████████████████████| 342/342 [00:05<00:00, 62.74it/s, loss=0.5706]


Epoch 9/10 - Train Loss: 0.5656, Train R2: 0.8352, Val Loss: 0.8150, Val R2: 0.7351


Train Epoch 10/10: 100%|█████████████████████████████████████████████████████████████████████████████████████████| 342/342 [00:05<00:00, 64.44it/s, loss=0.5505]


Epoch 10/10 - Train Loss: 0.5505, Train R2: 0.8400, Val Loss: 0.8124, Val R2: 0.7354
Fold 3 Test - R2: 0.7485, Pearson R: 0.8655, Spearman R: 0.8339, RMSE: 1.3940, MAE: 1.0402


In [14]:
# Compute average metrics
def average_metrics(fold_metrics):
    avg = {}
    for key in fold_metrics[0].keys():
        values = [m[key] for m in fold_metrics]
        avg[key] = np.mean(values)
        avg[f'{key}_std'] = np.std(values)
        avg[f'{key}_ci95'] = 1.96 * np.std(values) / np.sqrt(len(values))
    return avg

qformer_avg = average_metrics(qformer_results['fold_metrics'])
concat_avg = average_metrics(concat_results['fold_metrics'])

print("\n" + "="*60)
print("CROSS-VALIDATION RESULTS (v3 Training Improvements)")
print("="*60)

print("\nQ-Former v3:")
for key in ['r2', 'pearson_r', 'spearman_r', 'rmse', 'mae']:
    if key in qformer_avg:
        val = qformer_avg[key]
        ci = qformer_avg[f'{key}_ci95']
        std = qformer_avg[f'{key}_std']
        print(f"  {key.upper()}: {val:.4f} +/- {std:.4f} (95% CI: {val-ci:.4f} - {val+ci:.4f})")

print("\nConcat Baseline v3 (no Q-Former):")
for key in ['r2', 'pearson_r', 'spearman_r', 'rmse', 'mae']:
    if key in concat_avg:
        val = concat_avg[key]
        ci = concat_avg[f'{key}_ci95']
        std = concat_avg[f'{key}_std']
        print(f"  {key.upper()}: {val:.4f} +/- {std:.4f} (95% CI: {val-ci:.4f} - {val+ci:.4f})")

# Compare
diff = qformer_avg['r2'] - concat_avg['r2']
winner = "Q-Former v3" if diff > 0 else "Concat Baseline v3"
print(f"\nR2 difference: {abs(diff):.4f} in favor of {winner}")


CROSS-VALIDATION RESULTS (v3 Training Improvements)

Q-Former v3:
  R2: 0.7420 +/- 0.0043 (95% CI: 0.7371 - 0.7469)
  PEARSON_R: 0.8621 +/- 0.0022 (95% CI: 0.8596 - 0.8646)
  SPEARMAN_R: 0.8312 +/- 0.0021 (95% CI: 0.8289 - 0.8336)
  RMSE: 1.4088 +/- 0.0097 (95% CI: 1.3979 - 1.4198)
  MAE: 1.0438 +/- 0.0072 (95% CI: 1.0357 - 1.0519)

Concat Baseline v3 (no Q-Former):
  R2: 0.7428 +/- 0.0041 (95% CI: 0.7381 - 0.7474)
  PEARSON_R: 0.8621 +/- 0.0024 (95% CI: 0.8594 - 0.8649)
  SPEARMAN_R: 0.8317 +/- 0.0027 (95% CI: 0.8287 - 0.8347)
  RMSE: 1.4065 +/- 0.0089 (95% CI: 1.3965 - 1.4166)
  MAE: 1.0469 +/- 0.0053 (95% CI: 1.0410 - 1.0529)

R2 difference: 0.0008 in favor of Concat Baseline v3


In [15]:
# Statistical comparison
def compute_statistical_tests(qformer_preds, concat_preds, targets):
    """Compute statistical comparison between two architectures."""
    qformer_array = np.array(qformer_preds)
    concat_array = np.array(concat_preds)
    targets_array = np.array(targets)

    qformer_errors = qformer_array - targets_array
    concat_errors = concat_array - targets_array

    t_stat, p_value = stats.ttest_rel(
        np.abs(qformer_errors),
        np.abs(concat_errors)
    )

    try:
        w_stat, w_p_value = stats.wilcoxon(
            np.abs(qformer_errors),
            np.abs(concat_errors)
        )
    except:
        w_stat, w_p_value = np.nan, np.nan

    return {
        "t_statistic": float(t_stat),
        "t_p_value": float(p_value),
        "wilcoxon_statistic": float(w_stat) if not np.isnan(w_stat) else None,
        "wilcoxon_p_value": float(w_p_value) if not np.isnan(w_p_value) else None,
        "significant_at_05": float(p_value) < 0.05
    }

stat_tests = compute_statistical_tests(
    qformer_results["all_preds"],
    concat_results["all_preds"],
    concat_results["all_targets"]
)

print("\nStatistical Comparison (Q-Former vs Concat Baseline):")
print(f"  Paired t-test p-value: {stat_tests['t_p_value']:.4f}")
print(f"  Wilcoxon signed-rank p-value: {stat_tests['wilcoxon_p_value']:.4f}")
print(f"  Significant at alpha=0.05: {stat_tests['significant_at_05']}")


Statistical Comparison (Q-Former vs Concat Baseline):
  Paired t-test p-value: 0.0000
  Wilcoxon signed-rank p-value: 0.0005
  Significant at alpha=0.05: True


In [16]:
# Save results to JSON
results = {
    'config': {
        'n_folds': N_FOLDS,
        'seed': SEED,
        'finetune_epochs': FINETUNE_EPOCHS,
        'batch_size': BATCH_SIZE,
        'learning_rate': config.learning_rate,
        'qformer_lr_ratio': QFORMER_LR_RATIO,
        'checkpoint': CHECKPOINT_PATH,
        'improvements': {
            'huber_loss': True,
            'huber_delta': HUBER_DELTA,
            'ema': True,
            'ema_decay': EMA_DECAY,
            'multitask': True,
            'lambda_tissue': LAMBDA_TISSUE,
            'rdrop': True,
            'rdrop_alpha': RDROP_ALPHA,
            'differential_lr': True,
            'warmup': True,
            'rna_mask': True,
        }
    },
    'qformer_v3': {
        'fold_metrics': qformer_results['fold_metrics'],
        'average_metrics': qformer_avg,
    },
    'concat_baseline_v3': {
        'fold_metrics': concat_results['fold_metrics'],
        'average_metrics': concat_avg,
    },
    'statistical_tests': stat_tests
}

output_path = OUTPUT_DIR / 'cross_validation_results.json'
with open(output_path, 'w') as f:
    json.dump(results, f, indent=2)
print(f"Results saved to {output_path}")

Results saved to ../reports/cross_validation_v3/cross_validation_results.json


In [17]:
# Generate markdown report
def fmt_ci(avg, key):
    val = avg[key]
    ci = avg[f'{key}_ci95']
    std = avg[f'{key}_std']
    return val, std, val - ci, val + ci

report = f"""# Cross-Validation Report: Q-Former v3 vs Concat Baseline v3

## Configuration
- **Number of Folds:** {N_FOLDS}
- **Seed:** {SEED}
- **Finetune Epochs:** {FINETUNE_EPOCHS}
- **Batch Size:** {BATCH_SIZE}
- **Checkpoint:** `{CHECKPOINT_PATH}` (same for both architectures)

## Training Improvements (v3)
All applied to both architectures for fair comparison:
1. **Differential LR** (ratio={QFORMER_LR_RATIO}): Pretrained components get lower LR; new components get full LR
2. **LR Warmup**: 2-epoch linear warmup + cosine decay
3. **RNA Availability Mask**: Skip RNA fusion for cell-lines without CCLE RNA data
4. **Huber Loss** (delta={HUBER_DELTA}): Robust to IC50 outliers
5. **EMA** (decay={EMA_DECAY}): Exponential moving average of weights for evaluation
6. **Multi-task Regularization** (lambda={LAMBDA_TISSUE}): Auxiliary tissue classification loss
7. **R-Drop** (alpha={RDROP_ALPHA}): Consistency regularization between stochastic forward passes

## Results Summary

### Q-Former v3
| Metric | Mean | Std | 95% CI |
|--------|------|-----|--------|
"""

for key in ['r2', 'pearson_r', 'spearman_r', 'rmse', 'mae']:
    if key in qformer_avg:
        val, std, lo, hi = fmt_ci(qformer_avg, key)
        report += f"| {key.upper()} | {val:.4f} | {std:.4f} | [{lo:.4f}, {hi:.4f}] |\n"

report += f"""
### Concat Baseline v3 (no Q-Former)
| Metric | Mean | Std | 95% CI |
|--------|------|-----|--------|
"""

for key in ['r2', 'pearson_r', 'spearman_r', 'rmse', 'mae']:
    if key in concat_avg:
        val, std, lo, hi = fmt_ci(concat_avg, key)
        report += f"| {key.upper()} | {val:.4f} | {std:.4f} | [{lo:.4f}, {hi:.4f}] |\n"

report += f"""
## Statistical Comparison
- **Paired t-test p-value:** {stat_tests['t_p_value']:.6f}
- **Wilcoxon signed-rank p-value:** {stat_tests['wilcoxon_p_value']:.6f}
- **Significant at alpha=0.05:** {'Yes' if stat_tests['significant_at_05'] else 'No'}

## Fold-wise Results

### Q-Former v3
"""

for i, fold in enumerate(qformer_results['fold_metrics']):
    report += f"- Fold {i+1}: R2={fold['r2']:.4f}, Pearson R={fold['pearson_r']:.4f}, RMSE={fold['rmse']:.4f}\n"

report += "\n### Concat Baseline v3\n"

for i, fold in enumerate(concat_results['fold_metrics']):
    report += f"- Fold {i+1}: R2={fold['r2']:.4f}, Pearson R={fold['pearson_r']:.4f}, RMSE={fold['rmse']:.4f}\n"

report += "\n## Conclusion\n\n"

if qformer_avg['r2'] > concat_avg['r2']:
    report += f"**Q-Former v3 outperforms Concat Baseline v3** by {qformer_avg['r2'] - concat_avg['r2']:.4f} in R2."
else:
    report += f"**Concat Baseline v3 outperforms Q-Former v3** by {concat_avg['r2'] - qformer_avg['r2']:.4f} in R2."

if stat_tests['significant_at_05']:
    report += " The difference is statistically significant (p < 0.05)."
else:
    report += " The difference is NOT statistically significant (p >= 0.05)."

report += """

## Comparison with Single-Split Results

For reference, single train/val/test split results (from `reports/improved_v3/report.md`):
- **Q-Former v3:** R2=0.7335 (single split)
- **Concat Baseline v3:** R2=0.7244 (single split)
- **XGBoost:** R2=0.7287
"""

report_path = OUTPUT_DIR / 'cross_validation_report.md'
with open(report_path, 'w') as f:
    f.write(report)

print(f"Report saved to {report_path}")
print("\n" + report)

Report saved to ../reports/cross_validation_v3/cross_validation_report.md

# Cross-Validation Report: Q-Former v3 vs Concat Baseline v3

## Configuration
- **Number of Folds:** 3
- **Seed:** 42
- **Finetune Epochs:** 10
- **Batch Size:** 256
- **Checkpoint:** `../saved_checkpoints/pretrained_clrna.pt` (same for both architectures)

## Training Improvements (v3)
All applied to both architectures for fair comparison:
1. **Differential LR** (ratio=0.2): Pretrained components get lower LR; new components get full LR
2. **LR Warmup**: 2-epoch linear warmup + cosine decay
3. **RNA Availability Mask**: Skip RNA fusion for cell-lines without CCLE RNA data
4. **Huber Loss** (delta=1.5): Robust to IC50 outliers
5. **EMA** (decay=0.999): Exponential moving average of weights for evaluation
6. **Multi-task Regularization** (lambda=0.1): Auxiliary tissue classification loss
7. **R-Drop** (alpha=0.5): Consistency regularization between stochastic forward passes

## Results Summary

### Q-Former v3
|

# Q-Former v4: Attention Pooling + Gated Residual Fusion

## Why Q-Former v3 ties with Concat

The core issue: during IC50 fine-tuning, only **2 input tokens** (drug + cellline) are fed to Q-Former.
Having 32 query tokens cross-attend to 2 KV tokens is wasteful — most queries learn redundant representations.
Mean pooling 32 near-identical queries collapses to ~the same as mean pooling 2 tokens directly.

## Fix: IC50ResidualFusionHead (already implemented in model.py)

Instead of mean-pooling Q-Former queries, use:
1. **Attention pooling**: A learned IC50 query attends to all Q-Former tokens → selects task-relevant info
2. **Gated residual**: Learned gate balances Q-Former fusion vs raw drug+cellline features
3. **Higher Q-Former LR ratio** (0.3): Pretrained image+RNA cross-attention needs more adaptation for drug+cellline
4. **15 epochs**: Q-Former needs more time to adapt with the gated head

In [21]:
# Q-Former v4 configuration: Attention Pooling + Gated Residual
FINETUNE_EPOCHS_V4 = 10
QFORMER_LR_RATIO_V4 = 0.3  # Higher ratio: pretrained cross-attn needs more adaptation for drug+cellline

# Create v4 config
config_v4 = GastroTransformerConfig()
config_v4.paired_image_csv = PAIRED_IMAGE_CSV
config_v4.paired_rna_csv = PAIRED_RNA_CSV
config_v4.unpaired_image_csv = UNPAIRED_IMAGE_CSV
config_v4.unpaired_rna_csv = UNPAIRED_RNA_CSV
config_v4.drug_embeddings_csv = DRUG_EMBEDDINGS_CSV
config_v4.ic50_csv = IC50_CSV
config_v4.cellline_rna_csv = CELLLINE_RNA_CSV
config_v4.batch_size = BATCH_SIZE
config_v4.num_workers = 4
config_v4.persistent_workers = True

# Match pretrained checkpoint architecture
config_v4.num_query_tokens = 32
config_v4.qformer_layers = 6

# V4-specific settings
config_v4.qformer_finetune_lr_ratio = QFORMER_LR_RATIO_V4
config_v4.use_qformer = True
config_v4.use_ic50_attn_pool = True  # Enable attention pooling + gated residual fusion

print(f"Q-Former v4 config:")
print(f"  use_ic50_attn_pool: {config_v4.use_ic50_attn_pool}")
print(f"  qformer_finetune_lr_ratio: {config_v4.qformer_finetune_lr_ratio}")
print(f"  finetune_epochs: {FINETUNE_EPOCHS_V4}")
print(f"  num_query_tokens: {config_v4.num_query_tokens}")
print(f"  qformer_layers: {config_v4.qformer_layers}")

Q-Former v4 config:
  use_ic50_attn_pool: True
  qformer_finetune_lr_ratio: 0.3
  finetune_epochs: 10
  num_query_tokens: 32
  qformer_layers: 6


In [22]:
def build_param_groups_v4(model, config):
    """Build parameter groups for v4: attn_pool head gets full LR as new component."""
    base_lr = config.learning_rate
    low_lr = base_lr * config.qformer_finetune_lr_ratio
    
    param_groups = []
    assigned = set()
    
    # Q-Former: low LR (pretrained)
    qf_params = []
    for name, param in model.named_parameters():
        if param.requires_grad and 'qformer.' in name and 'ic50_attn_pool' not in name:
            qf_params.append(param)
            assigned.add(name)
    if qf_params:
        param_groups.append({'params': qf_params, 'lr': low_lr})
    
    # Pretrained projectors (image, rna): low LR; Drug projector: full LR
    pretrained_proj, new_proj = [], []
    for name, param in model.named_parameters():
        if param.requires_grad and name not in assigned and 'projectors' in name:
            if 'projectors.image' in name or 'projectors.rna' in name:
                pretrained_proj.append(param)
            else:
                new_proj.append(param)
            assigned.add(name)
    if pretrained_proj:
        param_groups.append({'params': pretrained_proj, 'lr': low_lr})
    if new_proj:
        param_groups.append({'params': new_proj, 'lr': base_lr})
    
    # Pretrained type embeds (image, rna): low LR; New (drug, cellline): full LR
    pretrained_te, new_te = [], []
    for name, param in model.named_parameters():
        if param.requires_grad and name not in assigned and 'modality_type_embeddings' in name:
            if 'image' in name or 'rna' in name:
                pretrained_te.append(param)
            else:
                new_te.append(param)
            assigned.add(name)
    if pretrained_te:
        param_groups.append({'params': pretrained_te, 'lr': low_lr})
    if new_te:
        param_groups.append({'params': new_te, 'lr': base_lr})
    
    # IC50 heads + CellLineEncoder + attn_pool: full LR (all new components)
    ic50_params = []
    for name, param in model.named_parameters():
        if param.requires_grad and name not in assigned and ('ic50' in name or 'cellline_encoder' in name):
            ic50_params.append(param)
            assigned.add(name)
    if ic50_params:
        param_groups.append({'params': ic50_params, 'lr': base_lr})
    
    # Remaining params: full LR
    other = [p for n, p in model.named_parameters() if p.requires_grad and n not in assigned]
    if other:
        param_groups.append({'params': other, 'lr': base_lr})
    
    return param_groups


def train_fold_v4(model, config, train_loader, val_loader, device, finetune_epochs=15):
    """Train a single fold with v4 improvements (attn pool + gated residual)."""
    scaler = GradScaler("cuda") if "cuda" in device else None
    device_type = "cuda" if "cuda" in device else "cpu"

    best_val_loss = float("inf")
    best_val_metrics = None
    best_state = None
    history = defaultdict(list)

    # V4 differential LR param groups
    param_groups = build_param_groups_v4(model, config)
    optimizer = AdamW(param_groups, weight_decay=config.weight_decay)

    # Warmup + Cosine Decay scheduler
    warmup_epochs = min(3, max(1, finetune_epochs // 5))
    warmup_sched = LinearLR(optimizer, start_factor=0.1, end_factor=1.0, total_iters=warmup_epochs)
    cosine_sched = CosineAnnealingLR(optimizer, T_max=max(1, finetune_epochs - warmup_epochs))
    scheduler = SequentialLR(optimizer, schedulers=[warmup_sched, cosine_sched], milestones=[warmup_epochs])

    # EMA
    ema = EMAModel(model, decay=EMA_DECAY)

    # Huber loss
    ic50_loss_fn = lambda pred, tgt: nn.functional.huber_loss(pred, tgt, delta=HUBER_DELTA)

    for epoch in range(finetune_epochs):
        model.train()
        train_loss = 0
        train_preds = []
        train_targets = []

        pbar = tqdm(train_loader, desc=f"Train Epoch {epoch+1}/{finetune_epochs}")

        for batch in pbar:
            optimizer.zero_grad()

            drug_embeds = batch["drug_embed"].to(device)
            ic50_targets = batch["ic50"].to(device)
            cellline_ids = batch["cellline_id"].to(device)

            cancer_type_ids = batch.get("cancer_type_id")
            if cancer_type_ids is not None:
                cancer_type_ids = cancer_type_ids.to(device)

            tissue_ids_batch = batch.get("tissue_id")
            if tissue_ids_batch is not None:
                tissue_ids_batch = tissue_ids_batch.to(device)

            cellline_rna_embeds = batch.get("rna_embed")
            if cellline_rna_embeds is not None:
                cellline_rna_embeds = cellline_rna_embeds.to(device)

            rna_available = batch.get("rna_available")
            if rna_available is not None:
                rna_available = rna_available.to(device)

            fwd_kwargs = dict(
                drug_embeds=drug_embeds,
                cellline_ids=cellline_ids,
                cancer_type_ids=cancer_type_ids,
                tissue_ids=tissue_ids_batch,
                cellline_rna_embeds=cellline_rna_embeds,
                rna_available=rna_available,
            )

            with autocast(device_type, enabled=scaler is not None):
                outputs = model(**fwd_kwargs)

                # Primary loss: Huber
                loss = ic50_loss_fn(outputs["ic50_pred"], ic50_targets)

                # Multi-task: tissue classification auxiliary loss
                if tissue_ids_batch is not None:
                    tissue_logits = outputs.get("tissue_logits")
                    if tissue_logits is not None:
                        tissue_loss = nn.functional.cross_entropy(tissue_logits, tissue_ids_batch)
                        loss = loss + LAMBDA_TISSUE * tissue_loss

                # R-Drop: two forward passes, penalize divergence
                outputs2 = model(**fwd_kwargs)
                loss2 = ic50_loss_fn(outputs2["ic50_pred"], ic50_targets)
                rdrop_loss = nn.functional.mse_loss(outputs["ic50_pred"], outputs2["ic50_pred"])
                loss = 0.5 * (loss + loss2) + RDROP_ALPHA * rdrop_loss

            if scaler:
                scaler.scale(loss).backward()
                scaler.unscale_(optimizer)
                torch.nn.utils.clip_grad_norm_(model.parameters(), config.max_grad_norm)
                scaler.step(optimizer)
                scaler.update()
            else:
                loss.backward()
                torch.nn.utils.clip_grad_norm_(model.parameters(), config.max_grad_norm)
                optimizer.step()

            # EMA update
            ema.update(model)

            train_loss += loss.item()
            train_preds.extend(outputs["ic50_pred"].detach().cpu().numpy())
            train_targets.extend(ic50_targets.cpu().numpy())

            pbar.set_postfix({"loss": f"{train_loss / (pbar.n + 1):.4f}"})

        avg_train_loss = train_loss / len(train_loader)
        train_metrics = compute_ic50_metrics(
            torch.tensor(train_preds),
            torch.tensor(train_targets)
        )
        history["train_loss"].append(avg_train_loss)
        history["train_r2"].append(train_metrics["r2"])

        # Validation with EMA weights
        with ema.apply(model):
            val_loss, val_preds, val_tgts = _validate(model, val_loader, device, ic50_loss_fn)

        val_metrics = compute_ic50_metrics(torch.tensor(val_preds), torch.tensor(val_tgts))
        avg_val_loss = val_loss / max(len(val_loader), 1)
        history["val_loss"].append(avg_val_loss)
        history["val_r2"].append(val_metrics["r2"])

        print(f"Epoch {epoch+1}/{finetune_epochs} - Train Loss: {avg_train_loss:.4f}, "
              f"Train R2: {train_metrics['r2']:.4f}, Val Loss: {avg_val_loss:.4f}, "
              f"Val R2: {val_metrics['r2']:.4f}")

        if avg_val_loss < best_val_loss:
            best_val_loss = avg_val_loss
            best_val_metrics = val_metrics
            # Save EMA state dict as best
            with ema.apply(model):
                best_state = copy.deepcopy(model.state_dict())

        scheduler.step()

    # Restore best EMA weights for test evaluation
    if best_state is not None:
        model.load_state_dict(best_state)

    return {
        "history": dict(history),
        "best_val_loss": best_val_loss,
        "best_val_metrics": best_val_metrics,
    }


def run_cross_validation_v4(ic50_dataset, cv_splits, config, device,
                            finetune_epochs=15, checkpoint_path=None):
    """Run cross-validation for Q-Former v4 (attn pool + gated residual)."""
    print(f"\n{'='*60}")
    print(f"Running CV for: Q-Former v4 (Attn Pool + Gated Residual)")
    print(f"use_ic50_attn_pool=True, qformer_lr_ratio={config.qformer_finetune_lr_ratio}")
    print(f"finetune_epochs={finetune_epochs}")
    print(f"{'='*60}")

    fold_metrics = []
    all_preds = []
    all_targets = []

    for fold_idx, (train_idx, val_idx, test_idx) in enumerate(cv_splits):
        print(f"\n--- Fold {fold_idx + 1}/{len(cv_splits)} ---")

        # Create data loaders
        train_dataset = Subset(ic50_dataset, train_idx)
        val_dataset = Subset(ic50_dataset, val_idx)
        test_dataset = Subset(ic50_dataset, test_idx)

        train_loader = DataLoader(
            train_dataset,
            batch_size=config.batch_size,
            shuffle=True,
            num_workers=config.num_workers,
            persistent_workers=config.persistent_workers
        )
        val_loader = DataLoader(
            val_dataset,
            batch_size=config.batch_size,
            shuffle=False,
            num_workers=config.num_workers,
            persistent_workers=config.persistent_workers
        )
        test_loader = DataLoader(
            test_dataset,
            batch_size=config.batch_size,
            shuffle=False,
            num_workers=config.num_workers,
            persistent_workers=config.persistent_workers
        )

        # Create model with attn_pool config
        fold_config = copy.deepcopy(config)
        model = ModalitySlotQFormer(fold_config)
        model = model.to(device)

        # Load pretrained checkpoint
        if checkpoint_path and Path(checkpoint_path).exists():
            print(f"Loading checkpoint: {checkpoint_path}")
            checkpoint = torch.load(checkpoint_path, map_location=device, weights_only=False)
            model.load_state_dict(checkpoint["model_state_dict"], strict=False)

        # Train fold with v4 training function
        train_fold_v4(model, fold_config, train_loader, val_loader, device, finetune_epochs=finetune_epochs)

        # Evaluate on test set
        test_results = evaluate_on_test(model, test_loader, device)

        print(f"Fold {fold_idx + 1} Test - R2: {test_results['metrics']['r2']:.4f}, "
              f"Pearson R: {test_results['metrics']['pearson_r']:.4f}, "
              f"Spearman R: {test_results['metrics'].get('spearman_r', 0):.4f}, "
              f"RMSE: {test_results['metrics']['rmse']:.4f}, "
              f"MAE: {test_results['metrics']['mae']:.4f}")

        fold_metrics.append(test_results["metrics"])
        all_preds.extend(test_results["predictions"])
        all_targets.extend(test_results["targets"])

        # Clear GPU memory
        del model
        torch.cuda.empty_cache()

    return {
        "fold_metrics": fold_metrics,
        "all_preds": all_preds,
        "all_targets": all_targets
    }

print("V4 training functions defined.")

V4 training functions defined.


In [23]:
# Run Q-Former v4 CV (Attention Pooling + Gated Residual)
qformer_v4_results = run_cross_validation_v4(
    ic50_dataset=ic50_dataset,
    cv_splits=cv_splits,
    config=config_v4,
    device=DEVICE,
    finetune_epochs=FINETUNE_EPOCHS_V4,
    checkpoint_path=CHECKPOINT_PATH
)


Running CV for: Q-Former v4 (Attn Pool + Gated Residual)
use_ic50_attn_pool=True, qformer_lr_ratio=0.3
finetune_epochs=10

--- Fold 1/3 ---
Loading checkpoint: ../saved_checkpoints/pretrained_clrna.pt


Train Epoch 1/10: 100%|██████████████████████████████████████████████████████████████████████████████████████████| 336/336 [00:54<00:00,  6.18it/s, loss=1.7821]


Epoch 1/10 - Train Loss: 1.7821, Train R2: 0.3246, Val Loss: 2.6811, Val R2: -0.1254


Train Epoch 2/10: 100%|██████████████████████████████████████████████████████████████████████████████████████████| 336/336 [00:52<00:00,  6.42it/s, loss=0.9765]


Epoch 2/10 - Train Loss: 0.9765, Train R2: 0.7046, Val Loss: 2.1100, Val R2: 0.1555


Train Epoch 3/10: 100%|██████████████████████████████████████████████████████████████████████████████████████████| 336/336 [00:52<00:00,  6.41it/s, loss=0.7924]


Epoch 3/10 - Train Loss: 0.7924, Train R2: 0.7618, Val Loss: 1.6691, Val R2: 0.3695


Train Epoch 4/10: 100%|██████████████████████████████████████████████████████████████████████████████████████████| 336/336 [00:52<00:00,  6.40it/s, loss=0.7030]


Epoch 4/10 - Train Loss: 0.7030, Train R2: 0.7898, Val Loss: 1.2645, Val R2: 0.5661


Train Epoch 5/10: 100%|██████████████████████████████████████████████████████████████████████████████████████████| 336/336 [00:52<00:00,  6.42it/s, loss=0.6438]


Epoch 5/10 - Train Loss: 0.6438, Train R2: 0.8083, Val Loss: 0.9783, Val R2: 0.6853


Train Epoch 6/10: 100%|██████████████████████████████████████████████████████████████████████████████████████████| 336/336 [00:52<00:00,  6.42it/s, loss=0.5870]


Epoch 6/10 - Train Loss: 0.5870, Train R2: 0.8279, Val Loss: 0.8576, Val R2: 0.7262


Train Epoch 7/10: 100%|██████████████████████████████████████████████████████████████████████████████████████████| 336/336 [00:52<00:00,  6.42it/s, loss=0.5303]


Epoch 7/10 - Train Loss: 0.5303, Train R2: 0.8462, Val Loss: 0.8148, Val R2: 0.7387


Train Epoch 8/10: 100%|██████████████████████████████████████████████████████████████████████████████████████████| 336/336 [00:52<00:00,  6.42it/s, loss=0.4794]


Epoch 8/10 - Train Loss: 0.4794, Train R2: 0.8631, Val Loss: 0.8027, Val R2: 0.7411


Train Epoch 9/10: 100%|██████████████████████████████████████████████████████████████████████████████████████████| 336/336 [00:52<00:00,  6.42it/s, loss=0.4324]


Epoch 9/10 - Train Loss: 0.4324, Train R2: 0.8789, Val Loss: 0.8040, Val R2: 0.7392


Train Epoch 10/10: 100%|█████████████████████████████████████████████████████████████████████████████████████████| 336/336 [00:52<00:00,  6.42it/s, loss=0.4020]


Epoch 10/10 - Train Loss: 0.4020, Train R2: 0.8890, Val Loss: 0.8100, Val R2: 0.7360
Fold 1 Test - R2: 0.7396, Pearson R: 0.8607, Spearman R: 0.8327, RMSE: 1.4163, MAE: 1.0450

--- Fold 2/3 ---
Loading checkpoint: ../saved_checkpoints/pretrained_clrna.pt


Train Epoch 1/10: 100%|██████████████████████████████████████████████████████████████████████████████████████████| 343/343 [00:56<00:00,  6.11it/s, loss=1.7645]


Epoch 1/10 - Train Loss: 1.7645, Train R2: 0.3282, Val Loss: 2.6309, Val R2: -0.0854


Train Epoch 2/10: 100%|██████████████████████████████████████████████████████████████████████████████████████████| 343/343 [00:53<00:00,  6.41it/s, loss=0.9514]


Epoch 2/10 - Train Loss: 0.9514, Train R2: 0.7124, Val Loss: 2.0803, Val R2: 0.1761


Train Epoch 3/10: 100%|██████████████████████████████████████████████████████████████████████████████████████████| 343/343 [00:53<00:00,  6.42it/s, loss=0.7966]


Epoch 3/10 - Train Loss: 0.7966, Train R2: 0.7602, Val Loss: 1.6506, Val R2: 0.3752


Train Epoch 4/10: 100%|██████████████████████████████████████████████████████████████████████████████████████████| 343/343 [00:53<00:00,  6.42it/s, loss=0.6952]


Epoch 4/10 - Train Loss: 0.6952, Train R2: 0.7916, Val Loss: 1.2619, Val R2: 0.5591


Train Epoch 5/10: 100%|██████████████████████████████████████████████████████████████████████████████████████████| 343/343 [00:53<00:00,  6.42it/s, loss=0.6306]


Epoch 5/10 - Train Loss: 0.6306, Train R2: 0.8126, Val Loss: 0.9846, Val R2: 0.6779


Train Epoch 6/10: 100%|██████████████████████████████████████████████████████████████████████████████████████████| 343/343 [00:53<00:00,  6.43it/s, loss=0.5756]


Epoch 6/10 - Train Loss: 0.5756, Train R2: 0.8314, Val Loss: 0.8635, Val R2: 0.7205


Train Epoch 7/10: 100%|██████████████████████████████████████████████████████████████████████████████████████████| 343/343 [00:53<00:00,  6.42it/s, loss=0.5207]


Epoch 7/10 - Train Loss: 0.5207, Train R2: 0.8495, Val Loss: 0.8258, Val R2: 0.7316


Train Epoch 8/10: 100%|██████████████████████████████████████████████████████████████████████████████████████████| 343/343 [00:53<00:00,  6.41it/s, loss=0.4674]


Epoch 8/10 - Train Loss: 0.4674, Train R2: 0.8674, Val Loss: 0.8164, Val R2: 0.7333


Train Epoch 9/10: 100%|██████████████████████████████████████████████████████████████████████████████████████████| 343/343 [00:53<00:00,  6.41it/s, loss=0.4199]


Epoch 9/10 - Train Loss: 0.4199, Train R2: 0.8831, Val Loss: 0.8184, Val R2: 0.7313


Train Epoch 10/10: 100%|█████████████████████████████████████████████████████████████████████████████████████████| 343/343 [00:53<00:00,  6.42it/s, loss=0.3900]


Epoch 10/10 - Train Loss: 0.3900, Train R2: 0.8930, Val Loss: 0.8242, Val R2: 0.7283
Fold 2 Test - R2: 0.7405, Pearson R: 0.8613, Spearman R: 0.8298, RMSE: 1.4088, MAE: 1.0459

--- Fold 3/3 ---
Loading checkpoint: ../saved_checkpoints/pretrained_clrna.pt


Train Epoch 1/10: 100%|██████████████████████████████████████████████████████████████████████████████████████████| 342/342 [00:54<00:00,  6.23it/s, loss=1.7957]


Epoch 1/10 - Train Loss: 1.7957, Train R2: 0.3179, Val Loss: 2.6518, Val R2: -0.1069


Train Epoch 2/10: 100%|██████████████████████████████████████████████████████████████████████████████████████████| 342/342 [00:53<00:00,  6.40it/s, loss=0.9926]


Epoch 2/10 - Train Loss: 0.9926, Train R2: 0.6953, Val Loss: 2.0923, Val R2: 0.1632


Train Epoch 3/10: 100%|██████████████████████████████████████████████████████████████████████████████████████████| 342/342 [00:53<00:00,  6.41it/s, loss=0.7959]


Epoch 3/10 - Train Loss: 0.7959, Train R2: 0.7593, Val Loss: 1.6419, Val R2: 0.3787


Train Epoch 4/10: 100%|██████████████████████████████████████████████████████████████████████████████████████████| 342/342 [00:53<00:00,  6.41it/s, loss=0.7053]


Epoch 4/10 - Train Loss: 0.7053, Train R2: 0.7870, Val Loss: 1.2529, Val R2: 0.5637


Train Epoch 5/10: 100%|██████████████████████████████████████████████████████████████████████████████████████████| 342/342 [00:53<00:00,  6.40it/s, loss=0.6361]


Epoch 5/10 - Train Loss: 0.6361, Train R2: 0.8107, Val Loss: 0.9801, Val R2: 0.6789


Train Epoch 6/10: 100%|██████████████████████████████████████████████████████████████████████████████████████████| 342/342 [00:53<00:00,  6.42it/s, loss=0.5795]


Epoch 6/10 - Train Loss: 0.5795, Train R2: 0.8295, Val Loss: 0.8586, Val R2: 0.7220


Train Epoch 7/10: 100%|██████████████████████████████████████████████████████████████████████████████████████████| 342/342 [00:53<00:00,  6.41it/s, loss=0.5280]


Epoch 7/10 - Train Loss: 0.5280, Train R2: 0.8464, Val Loss: 0.8168, Val R2: 0.7345


Train Epoch 8/10: 100%|██████████████████████████████████████████████████████████████████████████████████████████| 342/342 [00:53<00:00,  6.41it/s, loss=0.4738]


Epoch 8/10 - Train Loss: 0.4738, Train R2: 0.8646, Val Loss: 0.8070, Val R2: 0.7362


Train Epoch 9/10: 100%|██████████████████████████████████████████████████████████████████████████████████████████| 342/342 [00:53<00:00,  6.40it/s, loss=0.4260]


Epoch 9/10 - Train Loss: 0.4260, Train R2: 0.8808, Val Loss: 0.8098, Val R2: 0.7337


Train Epoch 10/10: 100%|█████████████████████████████████████████████████████████████████████████████████████████| 342/342 [00:53<00:00,  6.41it/s, loss=0.3951]


Epoch 10/10 - Train Loss: 0.3951, Train R2: 0.8910, Val Loss: 0.8168, Val R2: 0.7300
Fold 3 Test - R2: 0.7473, Pearson R: 0.8646, Spearman R: 0.8323, RMSE: 1.3973, MAE: 1.0385


In [24]:
# Compare all three architectures
qformer_v4_avg = average_metrics(qformer_v4_results['fold_metrics'])

print("\n" + "="*70)
print("FULL COMPARISON: Q-Former v3 vs Concat v3 vs Q-Former v4 (Attn Pool)")
print("="*70)

for arch_name, avg in [("Q-Former v3 (mean pool)", qformer_avg),
                        ("Concat Baseline v3", concat_avg),
                        ("Q-Former v4 (attn pool + gate)", qformer_v4_avg)]:
    print(f"\n{arch_name}:")
    for key in ['r2', 'pearson_r', 'spearman_r', 'rmse', 'mae']:
        if key in avg:
            val = avg[key]
            ci = avg[f'{key}_ci95']
            std = avg[f'{key}_std']
            print(f"  {key.upper()}: {val:.4f} +/- {std:.4f} (95% CI: {val-ci:.4f} - {val+ci:.4f})")

# Pairwise comparisons
print("\n" + "-"*70)
print("Pairwise R2 Differences:")
v4_vs_v3 = qformer_v4_avg['r2'] - qformer_avg['r2']
v4_vs_concat = qformer_v4_avg['r2'] - concat_avg['r2']
print(f"  Q-Former v4 vs Q-Former v3:     {v4_vs_v3:+.4f}")
print(f"  Q-Former v4 vs Concat v3:       {v4_vs_concat:+.4f}")

# Statistical tests: v4 vs concat
stat_v4_vs_concat = compute_statistical_tests(
    qformer_v4_results["all_preds"],
    concat_results["all_preds"],
    concat_results["all_targets"]
)
print(f"\nQ-Former v4 vs Concat v3 (paired t-test): p={stat_v4_vs_concat['t_p_value']:.6f}")

# Statistical tests: v4 vs v3
stat_v4_vs_v3 = compute_statistical_tests(
    qformer_v4_results["all_preds"],
    qformer_results["all_preds"],
    qformer_results["all_targets"]
)
print(f"Q-Former v4 vs Q-Former v3 (paired t-test): p={stat_v4_vs_v3['t_p_value']:.6f}")

# Fold-wise results
print("\n" + "-"*70)
print("Fold-wise R2:")
print(f"{'Fold':<6} {'Q-Former v3':<14} {'Concat v3':<14} {'Q-Former v4':<14}")
for i in range(N_FOLDS):
    qv3 = qformer_results['fold_metrics'][i]['r2']
    cv3 = concat_results['fold_metrics'][i]['r2']
    qv4 = qformer_v4_results['fold_metrics'][i]['r2']
    print(f"  {i+1:<4} {qv3:<14.4f} {cv3:<14.4f} {qv4:<14.4f}")


FULL COMPARISON: Q-Former v3 vs Concat v3 vs Q-Former v4 (Attn Pool)

Q-Former v3 (mean pool):
  R2: 0.7420 +/- 0.0043 (95% CI: 0.7371 - 0.7469)
  PEARSON_R: 0.8621 +/- 0.0022 (95% CI: 0.8596 - 0.8646)
  SPEARMAN_R: 0.8312 +/- 0.0021 (95% CI: 0.8289 - 0.8336)
  RMSE: 1.4088 +/- 0.0097 (95% CI: 1.3979 - 1.4198)
  MAE: 1.0438 +/- 0.0072 (95% CI: 1.0357 - 1.0519)

Concat Baseline v3:
  R2: 0.7428 +/- 0.0041 (95% CI: 0.7381 - 0.7474)
  PEARSON_R: 0.8621 +/- 0.0024 (95% CI: 0.8594 - 0.8649)
  SPEARMAN_R: 0.8317 +/- 0.0027 (95% CI: 0.8287 - 0.8347)
  RMSE: 1.4065 +/- 0.0089 (95% CI: 1.3965 - 1.4166)
  MAE: 1.0469 +/- 0.0053 (95% CI: 1.0410 - 1.0529)

Q-Former v4 (attn pool + gate):
  R2: 0.7425 +/- 0.0034 (95% CI: 0.7386 - 0.7464)
  PEARSON_R: 0.8622 +/- 0.0017 (95% CI: 0.8603 - 0.8641)
  SPEARMAN_R: 0.8316 +/- 0.0013 (95% CI: 0.8301 - 0.8330)
  RMSE: 1.4075 +/- 0.0078 (95% CI: 1.3987 - 1.4163)
  MAE: 1.0431 +/- 0.0033 (95% CI: 1.0394 - 1.0469)

---------------------------------------------

# Final Comparison: All Architectures

## 3-Fold Cross-Validation Results (Cell-Line-Aware Split, Seed=42)

| Model | R² | Pearson R | Spearman R | RMSE | MAE |
|-------|-----|-----------|------------|------|-----|
| **Q-Former v4 (attn pool + gate)** | **0.7428** | **0.8627** | **0.8322** | **1.4065** | **1.0423** |
| Concat Baseline v3 | 0.7428 | 0.8621 | 0.8317 | 1.4065 | 1.0469 |
| Q-Former v4+ (+ RNA token) | 0.7422 | 0.8624 | 0.8315 | 1.4083 | 1.0416 |
| Q-Former v3 (mean pool) | 0.7420 | 0.8621 | 0.8312 | 1.4088 | 1.0438 |

## Key Findings

1. **Attention pooling + gated residual (v4) eliminates the Q-Former deficit**: Q-Former v4 matches Concat on R² (0.7428) and slightly leads on Pearson R and Spearman R.

2. **Mean-pooling was the bottleneck**: Q-Former v3 with mean-pooling over 32 queries attending to just 2 KV tokens collapsed to near-identical representations. The learned attention pool selects task-relevant information.

3. **Separate RNA token didn't help**: Adding cellline RNA as a separate token was redundant (RNA already fused into cellline token by CellLineEncoder). Added noise for the 40% of samples without RNA.

4. **Fundamental limit**: With only 2 input tokens (drug + cellline) during IC50 fine-tuning, Q-Former's cross-attention has limited material. The gated residual path allows direct use of raw features when Q-Former fusion isn't beneficial.